firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -198,4 +215,48 @@ void shutdown() {
       Thread.currentThread().interrupt();
     }
   }
+
+  /**
+   * Sets the connectivity attempt timer to track. Existing timers must complete or be cancelled
+   * before a new timer can be set.
+   */
+  void setConnectivityAttemptTimer(DelayedTask connectivityAttemptTimer) {
+    // If an existing timer is already running, we want to continue using that one.
+    if (this.connectivityAttemptTimer != null) {
+      connectivityAttemptTimer.cancel();
+    } else {
+      try {
+        ManagedChannel channel = this.channelTask.getResult();
+        if (channel != null) {
+          this.connectivityAttemptTimer = connectivityAttemptTimer;
+          listenToConnectivityState(channel);
+        }
+      } catch (IllegalStateException e) {
+        // When the ManagedChannel is first initialized, channelTask could still be running.
+        connectivityAttemptTimer.cancel();
+      }
+    }
+  }
+
+  /**
+   * Cancels the connectivityStateTimer if the new state indicates grpc is online. Otherwise, we
+   * reset the listener for the next state change.
+   */
+  private void listenToConnectivityState(ManagedChannel channel) {
+    ConnectivityState newState = channel.getState(false);
+    // Check that the new state is online, then cancel timer.
+    if (newState == ConnectivityState.READY) {
+      clearConnectivityTimer();
+    } else {
+      channel.notifyWhenStateChanged(newState, () -> listenToConnectivityState(channel));
+    }
```

Comment:
Would it be possible for us to start the timer when the channel enters ConnectivityState.CONNECTING, and stop the timer when the channel enters ConnectivityState.READY?  And if the timer elapses can we somehow just "close" the channel (instead of marking it idle) in such a way that any pending streams / RPCs will just fail (in which case our existing retry behavior should kick in, but gRPC should use a new connection hopefully)?

Basically this would have two advantages:
1. It would be nice if this connection-timeout logic was 100% based on gRPC APIs (i.e. monitoring the ConnectivityState and resetting the channel when necessary) and it didn't bleed into the rest of our code (runBidiStreamingRpc, etc.).  That's cleaner and if/when gRPC improves their connection behavior, it would be trivial to just remove this timer logic (it could basically be a single method).
2. As-is, we'll only trigger this connection timeout if you're doing streaming RPCs.  If you're doing a non-streaming RPC (i.e. a transaction) then we'll still have the same problem with really long connection timeouts.

That said, I'm not sure how possible this is.  It looks like the two gRPC methods we might be able to play with are shutdownNow() [but then we'd presumably need to re-create the ManagedChannel?] or resetConnectBackoff() [but I don't know if that actually does anything useful in the 100% packet loss scenario].

On this note, if you haven't already, it would probably be worth adding some logging to listenToConnectivityState() and run your 100% packet loss scenario and verify what state(s) the gRPC connection goes through.  Is it just stuck CONNECTING the whole time?  Or does it do more than that?

Reply:
> Would it be possible for us to start the timer when the channel enters ConnectivityState.CONNECTING, and stop the timer when the channel enters ConnectivityState.READY? 

Yes, I will update to do so. It doesn't make much of a difference in practice in the current state though, since it shifts from IDLE to CONNECTING almost immediately. However, if we end up implementing the timer fully independent of the WatchStream, I can see why that makes sense.

> And if the timer elapses can we somehow just "close" the channel (instead of marking it idle) in such a way that any pending streams / RPCs will just fail (in which case our existing retry behavior should kick in, but gRPC should use a new connection hopefully)?

I originally thought about closing the underlying channel and calling `GrpcCallProvider.initChannel()` again, but that seemed like more overhead, since you would have to tear down and rebuild the channel each time. `enterIdle()` to reset the connection seemed like a cleaner approach that would only require initializing the channel a single time, rather than on each timeout. 

>On this note, if you haven't already, it would probably be worth adding some logging to listenToConnectivityState() and run your 100% packet loss scenario and verify what state(s) the gRPC connection goes through. Is it just stuck CONNECTING the whole time? Or does it do more than that?

From what I've observed, it starts in IDLE, immediately transitions to CONNECTING, and stays stuck in CONNECTING until it goes to TRANSIENT_FAILURE after a minute or two.


From here, I guess the main issue is whether or not to use `enterIdle()` or completely shutdown and reinitialize the channel. If we do a complete tear down, we can keep all the connectivity logic in the grpc provider. However, in the current state, if we don't tear the channel down, we need to re-run the runBidiStreamingRpc to trigger a new connection.

Reply:
This was marked resolved without describing the resolution.

Reply:
Resolution:

The connectivity timer starts when the channel enters a CONNECTING state, and if the timer elapses, the ManagedChannel will be torn down and reinitialized. However, if the channel enters any other state, the timer is cleared, and the state listener is set again.

Although it takes ~15-40ms to reset the channel, it's still a big enough win to justify the cost ([discussion](https://github.com/firebase/firebase-android-sdk/pull/1045#discussion_r356871348)).